File: orchestrator-api/src/main/java/com/sequenceiq/cloudbreak/orchestrator/model/ContainerConstraint.java
package com.sequenceiq.cloudbreak.orchestrator.model;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.sequenceiq.cloudbreak.orchestrator.model.port.TcpPortBinding;

public class ContainerConstraint {

    private final String[] cmd;

    private final Integer instances;

    private final List<Integer> ports;

    private final List<List<String>> constraints;

    private final Map<String, String> env;

    private final String networkMode;

    private final TcpPortBinding tcpPortBinding;

    private final ContainerName containerName;

    private final Map<String, String> links;

    private final String appName;

    private final List<String> hosts;

    private final Map<String, String> volumeBinds;

    private final Double cpu;

    private final Double mem;

    private final Double disk;

    private final String identifier;

    private ContainerConstraint(Builder builder) {
        cmd = builder.cmd;
        cpu = builder.cpus;
        mem = builder.mem;
        instances = builder.instances;
        ports = builder.ports;
        constraints = builder.constraints;
        volumeBinds = builder.volumeBinds;
        env = builder.env;
        networkMode = builder.networkMode;
        tcpPortBinding = builder.tcpPortBinding;
        hosts = builder.hosts;
        containerName = builder.containerName;
        links = builder.links;
        appName = builder.appName;
        disk = builder.disk;
        identifier = builder.identifier;
    }

    public String[] getCmd() {
        return cmd;
    }

    public Integer getInstances() {
        return instances;
    }

    public List<Integer> getPorts() {
        return ports;
    }

    public List<List<String>> getConstraints() {
        return constraints;
    }

    public Map<String, String> getVolumeBinds() {
        return volumeBinds;
    }

    public Map<String, String> getEnv() {
        return env;
    }

    public String getNetworkMode() {
        return networkMode;
    }

    public TcpPortBinding getTcpPortBinding() {
        return tcpPortBinding;
    }

    public List<String> getHosts() {
        return hosts;
    }

    public Double getCpu() {
        return cpu;
    }

    public Double getMem() {
        return mem;
    }

    public ContainerName getContainerName() {
        return containerName;
    }

    public Map<String, String> getLinks() {
        return links;
    }

    public String getAppName() {
        return appName;
    }

    public Double getDisk() {
        return disk;
    }

    public String getIdentifier() {
        return identifier;
    }

    public static class Builder {

        private String[] cmd;

        private List<Integer> ports = new ArrayList<>();

        private Double cpus;

        private Double mem;

        private Integer instances;

        private List<List<String>> constraints = new ArrayList<>();

        private Map<String, String> volumeBinds = new HashMap<>();

        private Map<String, String> env = new HashMap<>();

        private String networkMode;

        private TcpPortBinding tcpPortBinding;

        private List<String> hosts = new ArrayList<>();

        private ContainerName containerName;

        private Map<String, String> links = new HashMap<>();

        private String appName;

        private Double disk;

        private String identifier;

        public Builder containerConstraint(ContainerConstraint containerConstraint) {
            cmd = containerConstraint.getCmd();
            ports = containerConstraint.getPorts();
            cpus = containerConstraint.getCpu();
            mem = containerConstraint.getMem();
            instances = containerConstraint.getInstances();
            constraints = containerConstraint.getConstraints();
            volumeBinds = containerConstraint.getVolumeBinds();
            env = containerConstraint.getEnv();
            networkMode = containerConstraint.getNetworkMode();
            tcpPortBinding = containerConstraint.getTcpPortBinding();
            hosts = containerConstraint.getHosts();
            containerName = containerConstraint.getContainerName();
            links = containerConstraint.getLinks();
            appName = containerConstraint.getAppName();
            disk = containerConstraint.getDisk();
            identifier = containerConstraint.getIdentifier();
            return this;
        }

        public Builder cmd(String[] cmd) {
            this.cmd = cmd;
            return this;
        }

        public Builder ports(List<Integer> ports) {
            this.ports.addAll(ports);
            return this;
        }

        public Builder cpus(Double numberOfCpus) {
            cpus = numberOfCpus;
            return this;
        }

        public Builder memory(Double megaBytesOfMemory) {
            mem = megaBytesOfMemory;
            return this;
        }

        public Builder withDiskSize(Double diskSize) {
            disk = diskSize;
            return this;
        }

        public Builder instances(Integer numberOfInstances) {
            instances = numberOfInstances;
            return this;
        }

        public Builder constraints(List<List<String>> constraints) {
            this.constraints.addAll(constraints);
            return this;
        }

        public Builder addVolumeBindings(Map<String, String> volumeBinds) {
            this.volumeBinds.putAll(volumeBinds);
            return this;
        }

        public Builder addEnv(Map<String, String> env) {
            this.env.putAll(env);
            return this;
        }

        public Builder networkMode(String networkMode) {
            this.networkMode = networkMode;
            return this;
        }

        public Builder tcpPortBinding(TcpPortBinding binding) {
            tcpPortBinding = binding;
            return this;
        }

        public Builder addHosts(List<String> hosts) {
            this.hosts.addAll(hosts);
            return this;
        }

        public Builder withNamePrefix(String namePrefix) {
            containerName = new ContainerName(null, namePrefix);
            return this;
        }

        public Builder withName(String name) {
            containerName = new ContainerName(name, null);
            return this;
        }

        public Builder addLink(String hostContainerLink, String link) {
            links.put(hostContainerLink, link);
            return this;
        }

        public Builder withAppName(String appName) {
            this.appName = appName;
            return this;
        }

        public Builder withIdentifier(String identifier) {
            this.identifier = identifier;
            return this;
        }

        public ContainerConstraint build() {
            return new ContainerConstraint(this);
        }
    }
}

File: orchestrator-api/src/main/java/com/sequenceiq/cloudbreak/orchestrator/model/ContainerConstraint.kt
package com.sequenceiq.cloudbreak.orchestrator.model

import java.util.ArrayList
import java.util.HashMap

import com.sequenceiq.cloudbreak.orchestrator.model.port.TcpPortBinding

class ContainerConstraint private constructor(builder: ContainerConstraint.Builder) {

    val cmd: Array<String>
    val instances: Int?
    private val ports: List<Int>
    private val constraints: List<List<String>>
    private val env: Map<String, String>
    val networkMode: String
    val tcpPortBinding: TcpPortBinding
    val containerName: ContainerName
    private val links: Map<String, String>
    val appName: String

    private val hosts: List<String>
    private val volumeBinds: Map<String, String>
    val cpu: Double?
    val mem: Double?
    val disk: Double?


    init {
        this.cmd = builder.cmd
        this.cpu = builder.cpus
        this.mem = builder.mem
        this.instances = builder.instances
        this.ports = builder.ports
        this.constraints = builder.constraints
        this.volumeBinds = builder.volumeBinds
        this.env = builder.env
        this.networkMode = builder.networkMode
        this.tcpPortBinding = builder.tcpPortBinding
        this.hosts = builder.hosts
        this.containerName = builder.containerName
        this.links = builder.links
        this.appName = builder.appName
        this.disk = builder.disk
    }

    fun getPorts(): MutableList<Int> {
        return ports
    }

    fun getConstraints(): MutableList<List<String>> {
        return constraints
    }

    fun getVolumeBinds(): MutableMap<String, String> {
        return volumeBinds
    }

    fun getEnv(): MutableMap<String, String> {
        return env
    }

    fun getHosts(): MutableList<String> {
        return hosts
    }

    fun getLinks(): MutableMap<String, String> {
        return links
    }

    class Builder {

        private var cmd: Array<String>? = null
        private var ports: MutableList<Int> = ArrayList()
        private var cpus: Double? = null
        private var mem: Double? = null
        private var instances: Int? = null
        private var constraints: MutableList<List<String>> = ArrayList()
        private var volumeBinds: MutableMap<String, String> = HashMap()
        private var env: MutableMap<String, String> = HashMap()
        private var networkMode: String? = null
        private var tcpPortBinding: TcpPortBinding? = null
        private var hosts: MutableList<String> = ArrayList()
        private var containerName: ContainerName? = null
        private var links: MutableMap<String, String> = HashMap()
        private var appName: String? = null
        private var disk: Double? = null

        fun containerConstraint(containerConstraint: ContainerConstraint): Builder {
            this.cmd = containerConstraint.cmd
            this.ports = containerConstraint.getPorts()
            this.cpus = containerConstraint.cpu
            this.mem = containerConstraint.mem
            this.instances = containerConstraint.instances
            this.constraints = containerConstraint.getConstraints()
            this.volumeBinds = containerConstraint.getVolumeBinds()
            this.env = containerConstraint.getEnv()
            this.networkMode = containerConstraint.networkMode
            this.tcpPortBinding = containerConstraint.tcpPortBinding
            this.hosts = containerConstraint.getHosts()
            this.containerName = containerConstraint.containerName
            this.links = containerConstraint.getLinks()
            this.appName = containerConstraint.appName
            this.disk = containerConstraint.disk
            return this
        }

        fun cmd(cmd: Array<String>): Builder {
            this.cmd = cmd
            return this
        }

        fun ports(ports: List<Int>): Builder {
            this.ports.addAll(ports)
            return this
        }

        fun cpus(numberOfCpus: Double?): Builder {
            this.cpus = numberOfCpus
            return this
        }

        fun memory(megaBytesOfMemory: Double?): Builder {
            this.mem = megaBytesOfMemory
            return this
        }

        fun withDiskSize(diskSize: Double?): Builder {
            this.disk = diskSize
            return this
        }

        fun instances(numberOfInstances: Int?): Builder {
            this.instances = numberOfInstances
            return this
        }

        fun constraints(constraints: List<List<String>>): Builder {
            this.constraints.addAll(constraints)
            return this
        }

        fun addVolumeBindings(volumeBinds: Map<String, String>): Builder {
            this.volumeBinds.putAll(volumeBinds)
            return this
        }

        fun addEnv(env: Map<String, String>): Builder {
            this.env.putAll(env)
            return this
        }

        fun networkMode(networkMode: String): Builder {
            this.networkMode = networkMode
            return this
        }

        fun tcpPortBinding(binding: TcpPortBinding): Builder {
            this.tcpPortBinding = binding
            return this
        }

        fun addHosts(hosts: List<String>): Builder {
            this.hosts.addAll(hosts)
            return this
        }

        fun withNamePrefix(namePrefix: String): Builder {
            this.containerName = ContainerName(null, namePrefix)
            return this
        }

        fun withName(name: String): Builder {
            this.containerName = ContainerName(name, null)
            return this
        }

        fun addLink(hostContainerLink: String, link: String): Builder {
            this.links.put(hostContainerLink, link)
            return this
        }

        fun withAppName(appName: String): Builder {
            this.appName = appName
            return this
        }

        fun build(): ContainerConstraint {
            return ContainerConstraint(this)
        }
    }
}
